### PR TITLE
Add incident management and OONIRun CRUD

### PR DESCRIPTION
**Pull request by FedericoCeratto**
Add initial version of the incident management CRUD, the OONIRun CRUD and a small geolocation helper for the probes.
Also some refactoring. #676 
https://docs.google.com/document/d/1TdMBWD45j3bx7GRMIriMvey72viQeKrx7Ad6DyboLwg/

### CODE REVIEW

**Comment by FedericoCeratto** [comment]
Cherry-picking to master done as discussed.

**Comment by bassosimone** [comment]
Okay @FedericoCeratto, also https://github.com/ooni/backend/pull/678/commits/77759dad1f740b902b86a7ae5502a2fee174b017 is LGTM 🐳 and I'd say we can cherry-pick it as well. This leaves us with one last commit to review. I will continue my review tomorrow. Thank you!

**Review by bassosimone — COMMENTED** [review]
🐳 with minor comments

File: api/debian/changelog [hunk 1]
```diff
@@ -1,3 +1,9 @@
+ooni-api (1.0.58) unstable; urgency=medium
+
+  * Add incident management CRUD
```
[the remark this answers is not in the document]
Comment (FedericoCeratto) [reply api/debian/changelog:3]: It's ok, just a reminder. The real tracking of all the changes is in git.

File: api/ooniapi/incidents.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+                type: string
+            links:
+              type: array
+              items:
+                type: string
+    responses:
+      200:
+        type: object
+    """
+    """
```
[the remark this answers is not in the document]
Comment (FedericoCeratto) [reply api/ooniapi/incidents.py:242]: The block is commented out temporarily.

File: api/ooniapi/oonirun.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    get_account_id_or_none,
+)
+from ooniapi.config import metrics
+from ooniapi.database import query_click, optimize_table, insert_click, raw_query
+from ooniapi.errors import jerror
+from ooniapi.utils import nocachejson, cachedjson, generate_random_intuid
+
+log: logging.Logger
+
+# The table creation is in tests/integ/clickhouse_1_schema.sql
```
[the remark this answers is not in the document]
Comment (FedericoCeratto) [reply api/ooniapi/oonirun.py]: Done

File: api/ooniapi/incidents.py [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+log: logging.Logger
+
+# Support multi-node Clickhouse with ReplacingMergeTree
+# and update/delete without having to read the whole table before.
+# Updates/deletes are performed by ReplacingMergeTree based on update_time
+# against (title, event_type). We are not checking for the correct values
+# in the old dict during update/delete.
+# "deleted" will be used automatically by Clickhouse version 23.3
+#
+# The table creation is in tests/integ/clickhouse_1_schema.sql
```
[the remark this answers is not in the document]
Comment (FedericoCeratto) [reply api/ooniapi/incidents.py]: Done

File: api/ooniapi/incidents.py [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+    global log
+    log = current_app.logger
+    log.debug("showing incident")
+    try:
+        where = "WHERE id = %(id)s AND deleted != 1"
+        query_params = {"id": incident_id}
+        account_id = get_account_id_or_none()
+        if account_id is None:
+            # non-published incidents are not exposed to anon users
+            where += "\nAND published = 1"
```
[the remark this answers is not in the document]
Comment (FedericoCeratto) [reply api/ooniapi/incidents.py:122]: Before the AND? It's there to make the query a bit more readable in the logs.

File: api/ooniapi/incidents.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+        FROM incidents FINAL
+        {where}
+        LIMIT 1
+        """
+        q = query_click(query, query_params)
+        if len(q) < 1:
+            return jerror("Not found")
+        inc = q[0]
+        inc["published"] = bool(inc["published"])
+        return nocachejson(incident=inc, v=1)
```
[the remark this answers is not in the document]
Comment (FedericoCeratto) [reply api/ooniapi/incidents.py:136]: Comment updated, thanks!

**Comment by FedericoCeratto** [comment]
1.0.58~pr678-94  deployed on -test and FSN

**Comment by hellais** [comment]
This PR is also related to https://github.com/ooni/backend/issues/630

**Pull request merged**